### PR TITLE
 [fix] external_bang - UnicodeDecodeError: 'ascii' codec can't decode

### LINKED ISSUE #2041: Running "python searx/webapp.py" works, running via uwsgi doesn't
Hi,

I'm new to this, so I'm probably doing something really dumb. Apologies in advance if I'm time wasting.

I've installed Searx in a Ubuntu 20.4 Container (running under Proxmox)

All's gone well, except if I try to access my site, I get an Internal Server Error.

Looking into the /var/log/uwsgi/app/searx.log file I see

```
Sat Jul  4 17:19:29 2020 - setgid() to 998
Sat Jul  4 17:19:29 2020 - setuid() to 998
Sat Jul  4 17:19:29 2020 - Python version: 3.8.2 (default, Apr 27 2020, 15:53:34)  [GCC 9.3.0]
Sat Jul  4 17:19:29 2020 - PEP 405 virtualenv detected: /usr/local/searx/searx-pyenv
Sat Jul  4 17:19:29 2020 - Set PythonHome to /usr/local/searx/searx-pyenv
Sat Jul  4 17:19:29 2020 - Python main interpreter initialized at 0x55595f83fc40
Sat Jul  4 17:19:29 2020 - python threads support enabled
Sat Jul  4 17:19:29 2020 - your server socket listen backlog is limited to 100 connections
Sat Jul  4 17:19:29 2020 - your mercy for graceful operations on workers is 60 seconds
Sat Jul  4 17:19:29 2020 - mapped 218760 bytes (213 KB) for 2 cores
Sat Jul  4 17:19:29 2020 - *** Operational MODE: preforking ***
Sat Jul  4 17:19:29 2020 - added /usr/local/searx/searx-src/ to pythonpath.
Sat Jul  4 17:19:29 2020 - spawned uWSGI master process (pid: 24971)
Sat Jul  4 17:19:29 2020 - spawned uWSGI worker 1 (pid: 24977, cores: 1)
Sat Jul  4 17:19:29 2020 - spawned uWSGI worker 2 (pid: 24978, cores: 1)
Sat Jul  4 17:19:29 2020 - spawned uWSGI http 1 (pid: 24979)
Traceback (most recent call last):
  File "/usr/local/searx/searx-src/searx/webapp.py", line 74, in <module>
    from searx.search import SearchWithPlugins, get_search_query_from_webapp
  File "/usr/local/searx/searx-src/searx/search.py", line 32, in <module>
    from searx.external_bang import get_bang_url
  File "/usr/local/searx/searx-src/searx/external_bang.py", line 13, in <module>
    for bang in json.load(json_file)['bang']:
  File "/usr/lib/python3.8/json/__init__.py", line 293, in load
    return loads(fp.read(),
  File "/usr/lib/python3.8/encodings/ascii.py", line 26, in decode
    return codecs.ascii_decode(input, self.errors)[0]
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 31341: ordinal not in range(128)
Sat Jul  4 17:19:30 2020 - unable to load app 0 (mountpoint='') (callable not found or import error)
Sat Jul  4 17:19:30 2020 - *** no app loaded. going in full dynamic mode ***
Traceback (most recent call last):
  File "/usr/local/searx/searx-src/searx/webapp.py", line 74, in <module>
    from searx.search import SearchWithPlugins, get_search_query_from_webapp
  File "/usr/local/searx/searx-src/searx/search.py", line 32, in <module>
    from searx.external_bang import get_bang_url
  File "/usr/local/searx/searx-src/searx/external_bang.py", line 13, in <module>
    for bang in json.load(json_file)['bang']:
  File "/usr/lib/python3.8/json/__init__.py", line 293, in load
    return loads(fp.read(),
  File "/usr/lib/python3.8/encodings/ascii.py", line 26, in decode
    return codecs.ascii_decode(input, self.errors)[0]
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 31341: ordinal not in range(128)
Sat Jul  4 17:19:30 2020 - unable to load app 0 (mountpoint='') (callable not found or import error)
Sat Jul  4 17:19:30 2020 - *** no app loaded. going in full dynamic mode ***
Sat Jul  4 17:19:36 2020 - --- no python application found, check your startup logs for errors ---
Sat Jul  4 17:19:38 2020 - --- no python application found, check your startup logs for errors ---
Sat Jul  4 17:21:37 2020 - --- no python application found, check your startup logs for errors ---
```

The thing is, if I go through troubleshooting and do this:

```
# start webapp
$ sudo -H -u searx -i
(searx)$ cd /usr/local/searx/searx-src
(searx)$ export SEARX_SETTINGS_PATH=/etc/searx/settings.yml
(searx)$ python searx/webapp.py
```

searx works perfectly and I can browse/search and get results.

What have I done wrong thus that uwsgi can't serve the Python app?

Thanks!

### PR DESCRIPTION
**Pull request by return42**
Python's default encoding depends on the platform, set (python) default encoding
UTF-8 in uwsgi ini files:

    LANG=C.UTF-8
    LANGUAGE=C.UTF-8
    LC_ALL=C.UTF-8

Error pattern:

    Traceback (most recent call last):
      File "/usr/local/searx/searx-src/searx/webapp.py", line 74, in <module>
        from searx.search import SearchWithPlugins, get_search_query_from_webapp
      File "/usr/local/searx/searx-src/searx/search.py", line 32, in <module>
        from searx.external_bang import get_bang_url
      File "/usr/local/searx/searx-src/searx/external_bang.py", line 13, in <module>
        for bang in json.load(json_file)['bang']:
      File "/usr/lib/python3.8/json/__init__.py", line 293, in load
        return loads(fp.read(),
      File "/usr/lib/python3.8/encodings/ascii.py", line 26, in decode
        return codecs.ascii_decode(input, self.errors)[0]
      UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 31341: ordinal not in range(128)

close: #2041

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

### CODE REVIEW

**Comment by dalf** [comment]
There is this uwsgi.ini file too: ``` searx/dockerfiles/uwsgi.ini ``` ( [EDIT] but it is working as expected without modification...).

May be a belt and braces approach:
* replace this line:
https://github.com/asciimoo/searx/blob/4829a76aae8fa6af8d3f814ff780dc6e792b84e5/searx/external_bang.py#L12
*  with ```open(join(searx_dir, 'data/bangs.json'), encoding='utf-8') as json_file:```

So we make sure that even if the local is not utf-8, the file can still be read.

**Comment by return42** [comment]
> So we make sure that even if the local is not utf-8, the file can still be read.

Yes, in my very first attempt (a3de1abd) was a typo, I added the encoding in the wrong line :)  .. your example correct this.

> but it is working as expected without modification...

This depends on the (iso, container, docker ..) image . A common ubuntu image (and the alpine in our docker also) have [locale](https://man7.org/linux/man-pages/man7/locale.7.html) UTF-8.  But there are other images available which do not have a proper locale setup. By example; the ubuntu 20.04 container used by @tjharman in #2041 does not have UTF-8 locale setup / the alpine image has UTF-8.

> So we make sure that even if the local is not utf-8, the file can still be read.

Should we? For me, this is the central question we have to answer: do we support other encoding than UTF-8 ...

My answer is "NO" and setting [locale](https://man7.org/linux/man-pages/man7/locale.7.html) to UTF-8 is a  prerequisite of every searx process environment.

If we do otherwise .. we have to fix all file handles we open, not only the one in external_bang.py 

To summarize : we only support UTF-8 encoding.

[EDIT] should I add the `env` to `dockerfiles/uwsgi.ini` even if alpine has UTF-8 locale?  .. I would recommend, to clarify that we only support UTF-8. 

Thoughts?

**Comment by dalf** [comment]
```grep --include "*.py" --exclude-dir "local" -r "open("```


re-ordered output with some slight modifications to make it clear about the encoding:
```
# new one
searx/external_bang.py:with open(join(searx_dir, 'data/bangs.json')) as json_file:

# always use utf-8 encoding
searx/__init__.py:with open(settings_path, 'r', encoding='utf-8') as settings_yaml:
searx/engines/currency_convert.py:    json_data = open(current_dir + "/../data/currencies.json", 'r', encoding='utf-8').read()
searx/engines/__init__.py:languages = loads(open(engine_dir + '/../data/engines_languages.json', 'r', encoding='utf-8').read())
utils/fetch_languages.py:    with open(engines_languages_file, 'w', encoding='utf-8') as f:
searx/utils.py:useragents = json.loads(open(os.path.dirname(os.path.realpath(__file__)) + "/data/useragents.json", 'r', encoding='utf-8').read())

utils/fetch_currencies.py:f = open("currencies.json", "wb") ... json.dump(db, f, indent=4, encoding="utf-8")
utils/fetch_languages.py:    new_file = open(languages_file, 'wb');   ... new_file.write(file_content.encode('utf8'))

# unknown encoding (set according to the locale)
utils/fetch_firefox_version.py:with open(get_useragents_filename(), "w") as f:
setup.py:with open('README.rst') as f:
setup.py:with open('requirements.txt') as f:
setup.py:with open('requirements-dev.txt') as f:

```

It is messy, but the general pattern is (I've discovered / realized doing this grep):
* runtime the code is explicit about utf-8 encoding (whatever the environment configuration) .
* tools: it is more messy (```wb``` : the encoding is done elsewhere, about useragents.json can have ascii encoding without problem, 

I think the unwritten intent is to minimize the unexpected behavior and minimize the installation procedure: whatever the locale, it works.

But I think it is good to specify the UTF-8 (I think about some code in webapp.py dealing with date, babel, I don't know if the locale has an side-effect).

**Comment by return42** [comment]
> I think the unwritten intent is to minimize the unexpected behavior and minimize the installation procedure: whatever the locale, it works.

What weights more? I think _"minimize the unexpected behavior"_.  You can try to explicit set the encoding like you have shown, but you forget a lot of situations that are not covered by your _grep example_ / we have also jinja templates .. and all the pitfalls we don't think about right now.

Isn't it always the case that an application specifies its (default) encoding?

I vote for setting this default encoding when starting the application process.

According to my assessment: nowadays there are rarely images where the localization encoding is not UTF-8, mostly these are very small container images or images in small devices (IOT or other gadgets).

**Comment by dalf** [comment]
I haven't saw this edit:
> [EDIT] should I add the env to dockerfiles/uwsgi.ini even if alpine has UTF-8 locale? .. I would recommend, to clarify that we only support UTF-8.

where does alphine set the UTF-8 locale ? When I ```docker exec .... /bin/sh; su searx -; export``` I don't see the CHARSET, LANG variables ; but it still works.

The default encoding in Jinja (and Flask) is UTF8:
* https://jinja.palletsprojects.com/en/2.11.x/api/  ``` For Jinja the default encoding of templates is assumed to be utf-8.```
* https://github.com/pallets/jinja/blob/ca8b0b0287e320fe1f4a74f36910ef7ae3303d99/src/jinja2/loaders.py#L166
* https://github.com/pallets/flask/blob/024f0d384cf5bb65c76ac59f8ddce464b2dc2ca1/src/flask/helpers.py#L1016-L1023

My point was, despite the fact that encoding setting requirement is not documented, AFIK, we hadn't any encoding issue before ```searx/external_bang.py``` (Yes we had / have encoding issue in the engines, but that's a different topic).

**Review by dalf — APPROVED** [review]

**Comment by return42** [comment]
> I haven't saw this edit:

Sorry, my fail, I added this edit after your post.

> When I docker exec .... /bin/sh; su searx -; export

what does the command `locale` spits out?

> The default encoding in Jinja (and Flask) is UTF8:

good to know / thanks!

> we hadn't any encoding issue before .. engines, but that's a different topic

This is correct .. but: It is not possible for me to assess all side effects and which images have UTF-8 problems. May be this problem is limited to the image used in #2041? .. May be we have not made use of UTF-8 beside Jinja and Flask implementations? .. May be we have only [numeric character notation](https://en.wikipedia.org/wiki/Numeric_character_reference) used which is passed through to the client without any encoding/decoding on server side .....may be ... hard to say. 

Thanks for the reviewing!

**Comment by dalf** [comment]
> what does the command locale spits out?

```/bin/sh: locale: not found```

```
# ls /usr/share/
apk              ca-certificates  licenses         man              misc             terminfo
```

[There is no locale command](https://pkgs.alpinelinux.org/contents?file=locale&path=&name=&branch=v3.10) in [Alpine 3.10](https://github.com/asciimoo/searx/blob/master/Dockerfile#L1).

But there is Alpine 3.12... https://github.com/asciimoo/searx/pull/1983
